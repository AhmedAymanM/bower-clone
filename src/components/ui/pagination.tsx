import { useState, useEffect } from 'react'
import { getFormattedCount } from '@/utils/getFormattedCount'
import { cn, type ClassValue } from '@/utils/cn'
import { Button } from '@/components/ui/button'

type PaginationProps = {
  totalPages: number
  initialPage: number
  onPageChange?: (page: number) => void
  className?: ClassValue
}

export function Pagination({
  initialPage = 1,
  totalPages,
  onPageChange,
  className,
}: PaginationProps) {
  const [currentPage, setCurrentPage] = useState(initialPage)

  const onGetPreviousPage = () => setCurrentPage(currentPage - 1)
  const onGetNextPage = () => setCurrentPage(currentPage + 1)

  useEffect(() => {
    onPageChange?.(currentPage)
  }, [currentPage, onPageChange])

  return (
    <div className={cn('flex items-center gap-2', className)}>
      <Button
        disabled={!currentPage || currentPage === 1}
        onClick={onGetPreviousPage}
      >
        Previous
      </Button>
      <span className="font-bold">
        {getFormattedCount(currentPage, 'standard')} /{' '}
        {getFormattedCount(totalPages, 'standard')}
      </span>
      <Button disabled={currentPage === totalPages} onClick={onGetNextPage}>
        Next
      </Button>
    </div>
  )
}
