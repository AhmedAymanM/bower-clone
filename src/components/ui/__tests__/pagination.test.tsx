import { fireEvent, render } from '@testing-library/react'

import { Pagination } from '../pagination'

const commonProps = {
  selectedPage: 1,
  totalPages: 5,
}

describe('Pagination component', () => {
  test('snapshot matches', () => {
    const { asFragment } = render(<Pagination {...commonProps} />)

    expect(asFragment()).toMatchSnapshot()
  })
  // README: ideally the snapshot should be sufficient, but for the demo both
  // tests are provided
  test('renders Pagination component correctly', () => {
    const { getByLabelText, getByText } = render(
      <Pagination {...commonProps} />
    )

    const previousPageButton = getByLabelText(/previous page/i)
    const nextPageButton = getByLabelText(/next page/i)
    const pagesNumber = getByText(/1 \/ 5/i)

    expect(previousPageButton).toBeInTheDocument()
    expect(nextPageButton).toBeInTheDocument()
    expect(pagesNumber).toBeInTheDocument()
  })

  test('renders with custom className', () => {
    const { container } = render(
      <Pagination {...commonProps} className="test" />
    )

    expect(container.querySelector('.test')).toBeInTheDocument()
  })

  test('onPageChange is triggered on navigation interaction', () => {
    const onPageChange = vi.fn()

    const { getByLabelText } = render(
      <Pagination {...commonProps} onPageChange={onPageChange} />
    )

    expect(onPageChange).toHaveBeenCalled()
    const nextPageButton = getByLabelText(/next page/i)
    fireEvent.click(nextPageButton)
    expect(onPageChange).toHaveBeenCalled()
  })

  test('renders with disabled previous button if in first page', () => {
    const { getByLabelText } = render(<Pagination {...commonProps} />)

    const previousPageButton = getByLabelText(/previous page/i)
    const nextPageButton = getByLabelText(/next page/i)

    expect(previousPageButton).toBeDisabled()
    fireEvent.click(nextPageButton)
    expect(previousPageButton).not.toBeDisabled()
  })

  test('renders with disabled next button if in last page', () => {
    const { getByLabelText } = render(
      <Pagination {...commonProps} selectedPage={commonProps.totalPages} />
    )

    const previousPageButton = getByLabelText(/previous page/i)
    const nextPageButton = getByLabelText(/next page/i)

    expect(nextPageButton).toBeDisabled()
    fireEvent.click(previousPageButton)
    expect(nextPageButton).not.toBeDisabled()
  })
})
