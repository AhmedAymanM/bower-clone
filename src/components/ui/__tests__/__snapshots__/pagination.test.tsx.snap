// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination component > snapshot matches > 
      <DocumentFragment>
        <div
          class="flex items-center gap-2"
        >
          <button
            aria-label="Previous page"
            class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
            disabled=""
          >
            ⬅︎
          </button>
          <span
            class="font-bold"
          >
            1 / 5
          </span>
          <button
            aria-label="Next page"
            class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
          >
            ➡︎
          </button>
        </div>
      </DocumentFragment>
     1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-2"
  >
    <button
      aria-label="Previous page"
      class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
      disabled=""
    >
      ⬅︎
    </button>
    <span
      class="font-bold"
    >
      1 / 5
    </span>
    <button
      aria-label="Next page"
      class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
    >
      ➡︎
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination component > snapshot matches 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-2"
  >
    <button
      aria-label="Previous page"
      class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
      disabled=""
    >
      ⬅︎
    </button>
    <span
      class="font-bold"
    >
      1 / 5
    </span>
    <button
      aria-label="Next page"
      class="rounded-sm border border-solid border-gray p-2 text-lg font-bold text-blue-light hover:border-blue-light disabled:cursor-not-allowed disabled:border-gray/75 disabled:text-gray"
    >
      ➡︎
    </button>
  </div>
</DocumentFragment>
`;
